RPC connector for TigerBeetle

batiati opened the pull request:
Implements an input connector for streaming [TigerBeetle](https://tigerbeetle.com/) events via CDC.

The CDC process is stateless, so this connector requires a [`cache_resource`](https://docs.redpanda.com/redpanda-connect/components/caches/about) to persist progress and resume from the last acknowledged event when restarted.

- Includes documentation and integration tests.  
- Licensing information in the file header is pending.

For more information about TigerBeetle CDC, see https://docs.tigerbeetle.com/operating/cdc/

Jeffail commented on internal/impl/tigerbeetle/input_tigerbeetle.go:
```diff
@@ -0,0 +1,629 @@
+	return nil, nil, ctx.Err()
+}
+
+func newTigerbeetleInput(config *service.ParsedConfig, resources *service.Resources) (s service.BatchInput, err error) {
```
Thanks for setting this up. I believe we're happy to go ahead with a FOSS license so no need for license checks 👍

batiati commented:
Thanks for the review @Jeffail!

The `LintRule`s are really cool! I've implemented the validations and added unit tests for checking valid and invalid configurations.

Also, I replaced assertions that would `panic` with errors that will cause the connector to shut down.

Please let me know if I should add the license disclaimer to the source files. I didn't want to just copy from other files, as most say `Copyright 2024 ...`.

batiati commented on internal/impl/tigerbeetle/input_tigerbeetle.go:
```diff
@@ -0,0 +1,629 @@
+	return service.NewConfigSpec().
+		Beta().
+		Categories("Services").
+		Version("0.0.1").
```
The same for `Beta()`, when should we remove it?

Jeffail replied:
Yeah you can get rid of that, but it's also harmless to leave in.

mmatczuk commented on internal/impl/tigerbeetle/input_tigerbeetle.go:
```diff
@@ -0,0 +1,646 @@
+}
+
+// Extracts events from TigerBeetle.
+func (input *tigerbeetleInput) produce(ctx context.Context, client tb.Client, timestampLast uint64) error {
```
nit: how about using short variable name for self?

batiati replied:
I guess it would usually be named just `i` in Go!

I ended up using the [TigerStyle](https://github.com/tigerbeetle/tigerbeetle/blob/main/docs/TIGER_STYLE.md#naming-things) coding style, favoring short but meaningful names and avoiding single-letter variables and abbreviations (except for trivial counters/indexes and well-known abbreviations like `ctx`, `config`, etc).  

Although TigerStyle was originally designed for another programming language (Zig), I think we can apply most of the rules as long as they don't conflict with Go's idiomatic conventions. I don't want this code to feel alien 👽 compared with the rest of the code base!

It's your call, I'd be glad to follow your coding style otherwise.

batiati commented:
> Hi @batiati Is this PR ready for another re-review? 

Yes, it is ready for review. I've addressed all comments so far.

batiati commented:
CI is failing due to `CGO_ENABLED: 0`.

The TigerBeetle client requires CGO to consume the static library `lib_tbclient` (the underlying implementation of the VSR client). It includes pre-built libraries in-repo for all supported platforms.

But it seems there's a policy on builds that require external dependencies and CGO, like the ZeroMQ connector:

> By default Redpanda Connect does not build with components that require linking to external libraries. If you wish to build Redpanda Connect locally with this component then set the build tag x_benthos_extra:

josephwoodward commented on internal/impl/tigerbeetle/input_tigerbeetle.go:
```diff
@@ -0,0 +1,701 @@
+			for _, result := range results {
+				bytes, err := jsonSerialize(result)
+				if err != nil {
+					input.logger.Errorf("Unable to serialize as JSON: %w", err)
```
This should be %s as we're logging. That said, given the code is returning an error below which is eventually logged as part of the errorgroup, perhaps returning a wrapped error with the same error message would suffice to reduce double logging?

josephwoodward commented on internal/impl/tigerbeetle/input_tigerbeetle.go:
```diff
@@ -0,0 +1,701 @@
+	_ = idleTimer.Stop()
+
+	for {
+		err := input.checkRateLimit(ctx)
```
nitpick (non-blocking): perhaps a candidate for an inline error check.

josephwoodward commented on internal/impl/tigerbeetle/input_tigerbeetle.go:
```diff
@@ -0,0 +1,701 @@
+			// However, we wait for the *consumer* to begin flushing the current results
+			// before issuing a new query, avoiding unnecessary idle time for workloads
+			// with high frequency but low volume per batch.
+			rescheduled := idleTimer.Reset(input.config.idleInterval)
```
nitpick (non-blocking): another a candidate for an inline check.

josephwoodward commented on internal/impl/tigerbeetle/input_tigerbeetle.go:
```diff
@@ -0,0 +1,701 @@
+			}
+		}
+		if attempt == max_tries {
+			return fmt.Errorf("failed to access the rate limit after %d attemps", max_tries)
```
```suggestion
			return fmt.Errorf("failed to access the rate limit after %d attempts", max_tries)
```

josephwoodward commented on docs/modules/components/pages/inputs/tigerbeetle_cdc.adoc:
```diff
@@ -0,0 +1,200 @@
+component_type_dropdown::[]
+
+
+Enables TigerBeetle CDC streaming for RedPanda Connect.
```
```suggestion
Enables TigerBeetle CDC streaming for Redpanda Connect.
```

batiati commented:
> but for the short term we can't get rid of CGO.

Hi @Jeffail and @mmatczuk! I did some research to find alternatives, and we really can't remove the CGO dependency anytime soon. However, as we ship the client with pre-built static libraries for multiple targets, we don't depend too heavily on the system's libraries (as is the case with installing libzmq3-dev when the `x_benthos_extra` flag is present).

Please let me know what you suggest doing to teach the CI to build the TigerBeetle connector conditionally, and which pieces of documentation warnings we should include about it.

Jeffail commented:
Hey @batiati, we might try and enable CGO for some of our standard distributions of RPCN, but we almost definitely need to exclude this from our fips builds.

In order to proceed I think we should guard all of these files with a CGO build constraint so that non-cgo builds aren't broken. If you could update your PR to do this then we can merge and handle the rest in a follow-up PR, where we'll probably create fips builds with a cherry-picked list of connectors.
